pkg/option/option.go

```diff
@@ -115,200 +120,224 @@ func (om OptionMap) DeepCopy() OptionMap {
+	io.optsMU.RLock()
+	cpy := &IntOptions{
+		Opts:    io.Opts.DeepCopy(),
+		Library: io.Library,
 	}
-	bo.optsMU.RUnlock()
+	io.optsMU.RUnlock()
 	return cpy
 }
 
-func NewBoolOptions(lib *OptionLibrary) *BoolOptions {
-	return &BoolOptions{
+func NewIntOptions(lib *OptionLibrary) *IntOptions {
+	return &IntOptions{
 		Opts:    OptionMap{},
 		Library: lib,
 	}
 }
 
-func (bo *BoolOptions) isEnabled(key string) bool {
-	set, exists := bo.Opts[key]
-	return exists && set
+func (io *IntOptions) GetValue(key string) int {
+	value, exists := io.Opts[key]
+	if !exists {
+		return OptionDisabled
+	}
+	return value
+}
+
+func (io *IntOptions) IsEnabled(key string) bool {
+	io.optsMU.RLock()
+	defer io.optsMU.RUnlock()
+	return io.GetValue(key) != OptionDisabled
 }
 
-func (bo *BoolOptions) IsEnabled(key string) bool {
-	bo.optsMU.RLock()
-	defer bo.optsMU.RUnlock()
-	return bo.isEnabled(key)
+// SetValidated sets the option `key` to the specified value. The caller is
+// expected to have validated the input to this function.
+func (io *IntOptions) SetValidated(key string, value int) {
+	io.optsMU.Lock()
+	io.Opts[key] = value
+	io.optsMU.Unlock()
 }
 
-func (bo *BoolOptions) Set(key string, value bool) {
-	bo.optsMU.Lock()
-	bo.Opts[key] = value
-	bo.optsMU.Unlock()
+// SetBool sets the specified option to Enabled.
+func (io *IntOptions) SetBool(key string, value bool) {
+	intValue := OptionDisabled
+	if value {
+		intValue = OptionEnabled
+	}
+	io.optsMU.Lock()
+	io.Opts[key] = intValue
+	io.optsMU.Unlock()
 }
 
-func (bo *BoolOptions) Delete(key string) {
-	bo.optsMU.Lock()
-	delete(bo.Opts, key)
-	bo.optsMU.Unlock()
+func (io *IntOptions) Delete(key string) {
+	io.optsMU.Lock()
+	delete(io.Opts, key)
+	io.optsMU.Unlock()
 }
 
-func (bo *BoolOptions) SetIfUnset(key string, value bool) {
-	bo.optsMU.Lock()
-	if _, exists := bo.Opts[key]; !exists {
-		bo.Opts[key] = value
+func (io *IntOptions) SetIfUnset(key string, value int) {
+	io.optsMU.Lock()
+	if _, exists := io.Opts[key]; !exists {
+		io.Opts[key] = value
 	}
-	bo.optsMU.Unlock()
+	io.optsMU.Unlock()
 }
 
-func (bo *BoolOptions) InheritDefault(parent *BoolOptions, key string) {
-	bo.optsMU.RLock()
-	bo.Opts[key] = parent.isEnabled(key)
-	bo.optsMU.RUnlock()
+func (io *IntOptions) InheritDefault(parent *IntOptions, key string) {
+	io.optsMU.RLock()
+	io.Opts[key] = parent.GetValue(key)
+	io.optsMU.RUnlock()
 }
 
-func ParseOption(arg string, lib *OptionLibrary) (string, bool, error) {
-	enabled := true
+func ParseOption(arg string, lib *OptionLibrary) (string, int, error) {
+	result := OptionEnabled
 
 	if arg[0] == '!' {
-		enabled = false
+		result = OptionDisabled
 		arg = arg[1:]
 	}
 
 	optionSplit := strings.SplitN(arg, "=", 2)
 	arg = optionSplit[0]
 	if len(optionSplit) > 1 {
-		if !enabled {
-			return "", false, fmt.Errorf("Invalid boolean format")
+		if result == OptionDisabled {
+			return "", OptionDisabled, fmt.Errorf("Invalid boolean format")
```

Comment:
Is this error message still valid given that we use integers ?

Reply:
Yes. It's still valid to specify an option value as something like `!debug` and if you've instead specified `!debug=true`, it makes no sense and should be rejected as an invalid boolean option.